### PR TITLE
Support models larger than 2GB

### PR DESCRIPTION
**Pull request by tonypottera24**
### Description

1. Check model.ByteSize() to decide whether to use `infer_shapes_path` or `infer_shapes`.
2. Check model.ByteSize() to decide whether to use `save_as_external_data` on `onnx.save`.

Note that if `save_as_external_data=False`, the following `check_model` may fail.

### Motivation and Context

Current `extract_model` function does not support models larger than 2GB.
For example, issue https://github.com/onnx/onnx/issues/6368 is caused by large models like an UNet.
This PR improves the existing function by checking the size of the model.

### CODE REVIEW

File: onnx/utils.py [hunk 1]
```diff
@@ -234,12 +235,20 @@ def extract_model(
...
+            onnx.shape_inference.infer_shapes_path(input_path, output_path)
+            model = onnx.load(output_path)
+        else:
+            model = onnx.shape_inference.infer_shapes(model)
 
     e = Extractor(model)
     extracted = e.extract_model(input_names, output_names)
 
-    onnx.save(extracted, output_path)
+    if model.ByteSize() > MAXIMUM_PROTOBUF:
```
Comment (xadupre) [inline onnx/utils.py:247]: A model could be below MAXIMUM_PROTOBUF and above after shape inference. I would at least document this new behaviour wherever this makes sense.

File: onnx/utils.py [hunk 2]
```diff
@@ -234,12 +235,20 @@ def extract_model(
     model = onnx.load(input_path)
 
     if infer_shapes:
-        model = onnx.shape_inference.infer_shapes(model)
+        if model.ByteSize() > MAXIMUM_PROTOBUF:
```
[the remark this answers is not in the document]
Comment (tonypottera24) [reply onnx/utils.py]: Thank you @xadupre, I changed it to `os.path.getsize`.

File: onnx/utils.py [hunk 3]
```diff
@@ -10,6 +10,7 @@
 import onnx.helper
 import onnx.shape_inference
 from onnx import FunctionProto, ModelProto, NodeProto, TensorProto, ValueInfoProto
+from onnx.checker import MAXIMUM_PROTOBUF
```
Comment (justinchuby) [inline onnx/utils.py:13]: Import modules only

File: onnx/utils.py [hunk 4]
```diff
@@ -231,15 +232,22 @@ def extract_model(
...
+        model = onnx.load(input_path)
+        if infer_shapes:
+            model = onnx.shape_inference.infer_shapes(model)
 
     e = Extractor(model)
     extracted = e.extract_model(input_names, output_names)
 
-    onnx.save(extracted, output_path)
+    if extracted.ByteSize() > MAXIMUM_PROTOBUF:
+        location = os.path.basename(output_path) + "_data"
```
Comment (justinchuby) [inline onnx/utils.py:247]:
```suggestion
        location = os.path.basename(output_path) + ".data"
```

consider a .data extension? When the model extension is not onnx, for example json, the original will produce json_data etc which is a little arbitrary

File: onnx/utils.py (hunk 3 above)
Comment (justinchuby) [reply onnx/utils.py:13]: I would avoid importing a class or a function from a module.

Import only the module instead to keep the name space clean and provide readers with more context on where its members come from. For example, readers don’t need to backtrack to see `sleep` is a function from `time`, as opposed to a function defined in the file. https://google.github.io/styleguide/pyguide.html#22-imports

File: onnx/utils.py (hunk 4 above)
Comment (tonypottera24) [reply onnx/utils.py]: Thank you @justinchuby , I changed it to ".data"

File: onnx/utils.py (hunk 3 above)
Comment (tonypottera24) [reply onnx/utils.py]: Thank you @justinchuby , I changed it to `onnx.checker.MAXIMUM_PROTOBUF `

File: onnx/utils.py (hunk 1 above)
Comment (tonypottera24) [reply onnx/utils.py]: Added note
```
Note: When the extracted model size is larger than 2GB, the extra data will be saved in "output_path.data".
```

**Review by justinchuby — APPROVED** [review]

**Pull request merged**